# MNIST Digit Classification using PyTorch (CNN)

## Project Overview

This project implements a complete end-to-end multiclass image classification pipeline using a Convolutional Neural Network (CNN) in PyTorch on the MNIST handwritten digits dataset.

The objective is to design, train, validate, and evaluate a CNN that learns spatial features from grayscale digit images (0–9) while following standard deep learning and PyTorch best practices.

## Key Features

- Multiclass image classification (10 classes)
- CNN-based feature extraction
- Clean Train / Validation / Test split
- GPU acceleration (CUDA support)
- Cross-Entropy loss with raw logits
- Accuracy and loss tracking
- Model checkpointing
- Inference on unseen samples

## Dataset

- **Dataset:** MNIST Handwritten Digits  
- **Training samples:** 60,000  
- **Test samples:** 10,000  
- **Image size:** 28 × 28 (grayscale)

The training dataset is further split into:

- **Training set:** 85%
- **Validation set:** 15%

Input images are normalized to improve convergence and training stability.

## Model Architecture

The model is a Convolutional Neural Network (CNN) consisting of:

### Convolutional feature extractor:
- Conv2D → ReLU → MaxPooling  
- Conv2D → ReLU → MaxPooling  

### Fully connected classifier:
- Flatten  
- Linear → ReLU  
- Linear → Output (10 classes)

No softmax is applied inside the model.  
`CrossEntropyLoss` internally applies softmax during training.

## Training Strategy

- **Optimizer:** Adam  
- **Learning rate:** 0.001  
- **Loss function:** CrossEntropyLoss  
- **Batch size:** 64  
- **Epochs:** 5  
- **Evaluation metrics:** Loss and Accuracy  




In [ ]:
##importing dependencies
import torch
import torch.nn as nn
import torch.optim as optim

from torchvision.datasets import MNIST
from torchvision import transforms
from torch.utils.data import DataLoader, random_split

In [ ]:
# global seed for reproducibility
import torch
import numpy as np
import random

SEED = 42

torch.manual_seed(SEED)
torch.cuda.manual_seed(SEED)
torch.cuda.manual_seed_all(SEED)

np.random.seed(SEED)
random.seed(SEED)

torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

In [ ]:
##device selection
device = "cuda" if torch.cuda.is_available() else "cpu"
print("Using device:", device)



Using device: cuda


In [ ]:
##Dataset and dataloader
transform = transforms.Compose([
    transforms.ToTensor(),
    transforms.Normalize((0.1307,), (0.3081,))
])

full_train_data = MNIST(
    root="data",
    train=True,
    download=True,
    transform=transform
)

test_data = MNIST(
    root="data",
    train=False,
    download=True,
    transform=transform
)


100%|██████████| 9.91M/9.91M [00:02<00:00, 4.58MB/s]
100%|██████████| 28.9k/28.9k [00:00<00:00, 133kB/s]
100%|██████████| 1.65M/1.65M [00:01<00:00, 1.25MB/s]
100%|██████████| 4.54k/4.54k [00:00<00:00, 13.6MB/s]


In [ ]:
#  Train / Validation split
train_size = int(0.85 * len(full_train_data)) ##from 60000 , we are splitting train and val in 85: 15 ratio
val_size = len(full_train_data) - train_size

generator = torch.Generator().manual_seed(SEED)

train_data, val_data = random_split(
    full_train_data,
    [train_size, val_size],
    generator=generator
)

train_loader = DataLoader(
    train_data,
    batch_size=64,
    shuffle=True
)

val_loader = DataLoader(
    val_data,
    batch_size=64,
    shuffle=False
)

test_loader = DataLoader(
    test_data,
    batch_size=64,
    shuffle=False
)

In [ ]:
##Model Creation
class DigitModel(nn.Module):
    def __init__(self):
        super().__init__()
        self.features = nn.Sequential( # to learn patterns from image
            nn.Conv2d(1, 32, kernel_size=3, padding=1),  # [B, 32, 28, 28]
            nn.ReLU(),
            nn.MaxPool2d(2),                             # [B, 32, 14, 14]

            nn.Conv2d(32, 64, kernel_size=3, padding=1), # [B, 64, 14, 14]
            nn.ReLU(),
            nn.MaxPool2d(2)                              # [B, 64, 7, 7]
        )

        self.classifier = nn.Sequential( #to make final decision  (class prediction  ) based on learned patterns
            nn.Flatten(),
            nn.Linear(64 * 7 * 7, 128),
            nn.ReLU(),
            nn.Linear(128, 10)
        )

    def forward(self, x):
        x = self.features(x)
        x = self.classifier(x)
        return x
model = DigitModel().to(device)


In [ ]:
##Loss and optimizer
loss_fn = nn.CrossEntropyLoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

In [ ]:
##training and validation
epochs = 5
best_val_acc = 0.0

for epoch in range(epochs):

    # -------- TRAIN --------
    model.train()
    train_loss = 0.0
    train_correct = 0
    train_total = 0

    for xb, yb in train_loader:
        xb = xb.to(device)
        yb = yb.to(device)

        optimizer.zero_grad()
        logits = model(xb)
        loss = loss_fn(logits, yb)
        loss.backward()
        optimizer.step()

        train_loss += loss.item()
        preds = torch.argmax(logits, dim=1)
        train_correct += (preds == yb).sum().item()
        train_total += yb.size(0)

    avg_train_loss = train_loss / len(train_loader)
    train_acc = train_correct / train_total

    # -------- VALIDATION --------
    model.eval()
    val_loss = 0.0
    val_correct = 0
    val_total = 0

    with torch.no_grad():
        for xb, yb in val_loader:
            xb = xb.to(device)
            yb = yb.to(device)

            logits = model(xb)
            loss = loss_fn(logits, yb)

            val_loss += loss.item()
            preds = torch.argmax(logits, dim=1)
            val_correct += (preds == yb).sum().item()
            val_total += yb.size(0)

    avg_val_loss = val_loss / len(val_loader)
    val_acc = val_correct / val_total

    # Saving the best model
    if val_acc > best_val_acc:
        best_val_acc = val_acc
        torch.save(model.state_dict(), "mnist_cnn_best.pth")
        print(f" Best model saved at epoch {epoch+1}")

    # -------- LOGS --------
    print(
        f"Epoch {epoch+1:03d} | "
        f"Train Loss: {avg_train_loss:.4f} | "
        f"Train Acc: {train_acc:.4f} | "
        f"Val Loss: {avg_val_loss:.4f} | "
        f"Val Acc: {val_acc:.4f}"
    )

 Best model saved at epoch 1
Epoch 001 | Train Loss: 0.1423 | Train Acc: 0.9562 | Val Loss: 0.0621 | Val Acc: 0.9826
 Best model saved at epoch 2
Epoch 002 | Train Loss: 0.0438 | Train Acc: 0.9862 | Val Loss: 0.0426 | Val Acc: 0.9873
Epoch 003 | Train Loss: 0.0294 | Train Acc: 0.9906 | Val Loss: 0.0408 | Val Acc: 0.9873
 Best model saved at epoch 4
Epoch 004 | Train Loss: 0.0210 | Train Acc: 0.9932 | Val Loss: 0.0426 | Val Acc: 0.9884
Epoch 005 | Train Loss: 0.0161 | Train Acc: 0.9944 | Val Loss: 0.0464 | Val Acc: 0.9872


## Model Training Results

The CNN model was trained for **5 epochs** with validation monitoring and checkpointing enabled.

### Final Performance Metrics

- **Best Validation Accuracy:** 98.84% (Epoch 4)
- **Final Training Accuracy:** 99.44%
- **Final Validation Accuracy:** 98.72%
- **Lowest Validation Loss:** 0.0408 (Epoch 3)

### Observations

- The model converged rapidly within the first 3 epochs.
- Validation accuracy remained consistently above 98.7%, indicating strong generalization.
- The gap between training and validation accuracy is minimal, suggesting no significant overfitting.
- Best model weights were saved based on validation performance.

In [ ]:
## reloading  the best model for test evaluation and inference
model.load_state_dict(
    torch.load("mnist_cnn_best.pth", map_location=device)
)
print(" Best model loaded for inference.")

 Best model loaded for inference.


In [ ]:
## Final Test Evaluation
model.eval()
test_loss = 0.0
test_correct = 0
test_total = 0

with torch.no_grad():
    for xb, yb in test_loader:
        xb = xb.to(device)
        yb = yb.to(device)

        logits = model(xb)
        loss = loss_fn(logits, yb)

        test_loss += loss.item()
        preds = torch.argmax(logits, dim=1)
        test_correct += (preds == yb).sum().item()
        test_total += yb.size(0)

print(
    f"Final Test Loss: {test_loss / len(test_loader):.4f} | "
    f"Test Acc: {test_correct / test_total:.4f}"
)


Final Test Loss: 0.0308 | Test Acc: 0.9911


## Final Test Evaluation

The trained CNN model was evaluated on the held-out test dataset to measure its generalization performance on unseen data.

### Evaluation Procedure

- The model was switched to evaluation mode using `model.eval()` to disable dropout and batch normalization updates.
- Gradient computation was disabled using `torch.no_grad()` to improve inference efficiency and reduce memory usage.
- The test dataset was passed through the model in batches.
- Test loss was computed using the defined loss function.
- Predictions were obtained using `argmax` on the output logits.
- Accuracy was calculated by comparing predicted labels with true labels.

### Metrics Reported

- **Test Loss:** Average loss across all test batches.
- **Test Accuracy:** Ratio of correctly predicted samples to total samples.

### Purpose

This evaluation provides an unbiased estimate of the model’s real-world performance, as the test dataset was not used during training or validation.

In [ ]:
##inference function
def predict(model, image_tensor, device, return_probs=False):
    """
    image_tensor: torch.Tensor of shape [1, 1, 28, 28]
    """
    model.eval()
    with torch.no_grad():
        logits = model(image_tensor.to(device))
        probs = torch.softmax(logits, dim=1)
        pred = probs.argmax(dim=1).item()

    if return_probs:
        return pred, probs
    return pred

In [ ]:
##inference on test data
sample, _ = test_data[0]          # [1, 28, 28]
sample = sample.unsqueeze(0)      # [1, 1, 28, 28]

pred, probs = predict(model, sample, device, return_probs=True)

print("Probabilities:", probs)
print("Predicted digit:", pred)

Probabilities: tensor([[3.6227e-10, 9.7889e-10, 1.4100e-09, 8.0887e-07, 1.1386e-08, 6.8217e-10,
         1.9798e-14, 1.0000e+00, 6.4846e-09, 1.2724e-07]], device='cuda:0')
Predicted digit: 7
